repair/repair.cc

```diff
@@ -236,16 +236,26 @@ tracker::~tracker() {
 
 void tracker::start(int id) {
     _gate.enter();
-    _status[id] = repair_status::RUNNING;
+    try {
+        _status[id] = repair_status::RUNNING;
+    } catch (...) {
+        _gate.leave();
+        throw;
+    }
 }
 
 void tracker::done(int id, bool succeeded) {
-    if (succeeded) {
-        _status.erase(id);
-    } else {
-        _status[id] = repair_status::FAILED;
+    try {
+        if (succeeded) {
+            _status.erase(id);
+        } else {
+            _status[id] = repair_status::FAILED;
+        }
+        _gate.leave();
```

Comment:
It's definitely better but still not perfect.
What if `tracker::done(id)` is called without the preceding successful `tracker::start(id)` call?
In this case there still going to be unbalanced `gate::enter()` + `gate::leave()` calls amount.

The same is if `start(id)` is called twice for the same `id`...

You need to check that `_status[id]` exists before going forward both in `start(id)` and in `done(id)`.

Reply:
I think it is the job of the caller to guarantee it will have balanced call of start() and done(). It is like the gate won't protect you from calling unbalanced enter and leave().

Checking if _status[id] won't help for example, in done() it goes to the branch that does not erase the id. If we call done() multiple times, we will call gate::leave() more than gate::enter().

Reply:
Gate has `with_gate()` to ensure that. 
gate::enter() and leave() are very low level methods and one indeed should be very careful when calling them and this low level API is indeed very unsafe - hence there is `with_gate()`

Ensuring safety with `tracker` *is* possible. Since it's much more primitive abstraction than the `gate` you can even make it safe without needing to implement `with_xxx()`. All what's needed is to separate the "active" state from the "status".

You can go with the `with_xxx` semantics of course - which will be MUCH safer than the split semantics you have now.

Reply:
Currently it use `defter` to make sure gate.leave() is called when the defter is destroyed. But, yes, I think with_gate is better here. I will send a new version shortly.

Reply:
> Currently it use `defter` to make sure gate.leave() is called when the defter is destroyed. But, yes, I think with_gate is better here. I will send a new version shortly.

I meant to implement something like `with_tracker(id, callable)`. This will make sure the above complications are not needed.

You implemented a split API in the tracker and it always leaves a place of a misuse (as you and I mentioned before). You can still make it quite safe even with the current API as I described above however if you implement a `with_tracker(...)` single-call notation this will be a much safer API.

Another small benefit is going to be that you would be able to easily incorporate a `with_gate(...)` into a `with_tracker(...)` API implementation.

Reply:
See my latest patch. It ready has a with_gate. with_tracker only wraps with_gate, I do not think it is very useful.